extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/RecordableBootstrapFactory.java

```diff
@@ -0,0 +1,63 @@
+package io.quarkus.hibernate.orm.runtime.boot;
+
+import org.hibernate.boot.registry.BootstrapServiceRegistry;
+import org.hibernate.boot.registry.classloading.spi.ClassLoaderService;
+import org.hibernate.boot.registry.internal.BootstrapServiceRegistryImpl;
+import org.hibernate.boot.registry.selector.spi.StrategySelector;
+
+import io.quarkus.hibernate.orm.runtime.customized.QuarkusIntegratorServiceImpl;
+import io.quarkus.hibernate.orm.runtime.customized.QuarkusStrategySelectorBuilder;
+import io.quarkus.hibernate.orm.runtime.recording.RecordableBootstrap;
+import io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService;
+import io.quarkus.hibernate.orm.runtime.service.InitialInitiatorListProvider;
+import io.quarkus.hibernate.orm.runtime.service.StandardHibernateORMInitiatorListProvider;
+
+final class RecordableBootstrapFactory {
+
+    private static final InitialInitiatorListProvider reactiveInitiatorListProvider = initReactiveListProviderMaybe();
+    private static final InitialInitiatorListProvider classicInitiatorListProvider = new StandardHibernateORMInitiatorListProvider();
+
+    public static RecordableBootstrap createRecordableBootstrapBuilder(QuarkusPersistenceUnitDefinition puDefinition) {
+        final BootstrapServiceRegistry bsr = buildBootstrapServiceRegistry();
+        final RecordableBootstrap ssrBuilder = new RecordableBootstrap(bsr, getInitiatorListProvider(puDefinition));
+        return ssrBuilder;
+    }
+
+    private static InitialInitiatorListProvider getInitiatorListProvider(QuarkusPersistenceUnitDefinition puDefinition) {
+        if (puDefinition.isReactive()) {
+            if (reactiveInitiatorListProvider == null) {
+                throw new IllegalStateException(
+                        "InitiatorList requires for Hibernate Reactive but Hibernate Reactive extension is not around?");
+            }
+            return reactiveInitiatorListProvider;
+        } else {
+            return classicInitiatorListProvider;
+        }
+    }
+
+    private static BootstrapServiceRegistry buildBootstrapServiceRegistry() {
+        final ClassLoaderService providedClassLoaderService = FlatClassLoaderService.INSTANCE;
+        // N.B. support for custom IntegratorProvider injected via Properties (as
+        // instance) removed
+
+        final QuarkusIntegratorServiceImpl integratorService = new QuarkusIntegratorServiceImpl(providedClassLoaderService);
+        final QuarkusStrategySelectorBuilder strategySelectorBuilder = new QuarkusStrategySelectorBuilder();
+        final StrategySelector strategySelector = strategySelectorBuilder.buildSelector(providedClassLoaderService);
+        return new BootstrapServiceRegistryImpl(true, providedClassLoaderService, strategySelector, integratorService);
+    }
+
+    private static InitialInitiatorListProvider initReactiveListProviderMaybe() {
```

Comment:
In ORM Panache I have used a list of `BuildItem` to register optional stuff like these. Otherwise a service loader would work too. I guess both would be better than reflection, no?

Reply:
I considered that, but this code is invoked from within a Bytecode Recorder.

This implies that if I were to pass the full list of Initiators to the recorder from the layer in which I have BuildItems, I'd need to load and proxy many more classes - which then get loaded by reflection anyway.

If I were to "just" pass the InitialInitiatorListProvider it would also need to be recorded, proxied and loaded again.. with the added complexity of having a BuildItem needed for this, and need some way to enforce it to be one class at most?

Happy to revisit, but I don't feel like it would be much better.

Reply:
P.B. I considered the ServiceLoader at first, but I really want "only one" - if there's a list of such providers it's a problem, it's not meant as an SPI. Single-named constant seems more fitting.

Reply:
Fair enough :)